pkg/driver/provisioner_server_impl.go

```
@@ -44,6 +45,92 @@ type ProvisionerServer struct {
+		grpcCode:      codes.InvalidArgument,
+		logMessage:    "Invalid bucket name - will not retry",
+	},
+	{
+		errorPatterns: []string{"InvalidLocationConstraint"},
+		grpcCode:      codes.InvalidArgument,
+		logMessage:    "Invalid location constraint - will not retry",
+	},
+	{
+		errorPatterns: []string{"AccessDenied"},
+		grpcCode:      codes.PermissionDenied,
+		logMessage:    "Access denied - will not retry",
+	},
+	{
+		errorPatterns: []string{"InvalidRequest", "MalformedXML"},
+		grpcCode:      codes.InvalidArgument,
+		logMessage:    "Invalid request format - will not retry",
+	},
+}
+
+// deleteBucketErrorMappings defines error mappings for bucket deletion
+var deleteBucketErrorMappings = []errorMapping{
+	{
+		errorPatterns: []string{"BucketNotEmpty"},
+		grpcCode:      codes.FailedPrecondition,
+		logMessage:    "Cannot delete non-empty bucket - will not retry",
+	},
+	{
+		errorPatterns: []string{"AccessDenied", "Forbidden"},
+		grpcCode:      codes.PermissionDenied,
+		logMessage:    "Access denied for bucket deletion - will not retry",
+	},
+	{
+		errorPatterns: []string{"InvalidBucketName", "InvalidRequest", "MalformedXML"},
+		grpcCode:      codes.InvalidArgument,
+		logMessage:    "Invalid request for bucket deletion - will not retry",
+	},
+}
+
+// mapErrorToGRPCStatus maps an error to appropriate gRPC status based on error mappings
+func mapErrorToGRPCStatus(err error, mappings []errorMapping, bucketName string) *status.Status {
+	errMessage := err.Error()
+
+	for _, mapping := range mappings {
+		for _, pattern := range mapping.errorPatterns {
+			if strings.Contains(errMessage, pattern) {
+				klog.ErrorS(err, mapping.logMessage, "bucketName", bucketName)
+				return status.New(mapping.grpcCode, errMessage)
+			}
+		}
+	}
+
+	// Default to Internal error for unmapped errors (will be retried)
+	klog.ErrorS(err, "Unmapped error - will retry with backoff", "bucketName", bucketName)
+	return status.New(codes.Internal, errMessage)
```

I think `errMessage` is never logged nor wrapped in the final error,  so the original error is lost